// purify-ts/Maybe.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Nothing = exports.Just = exports.Maybe = void 0;
const Either_js_1 = require("./Either.js");
exports.Maybe = {
    of(value) {
        return just(value);
    },
    empty() {
        return nothing;
    },
    zero() {
        return nothing;
    },
    fromNullable(value) {
        return value == null ? nothing : just(value);
    },
    fromFalsy(value) {
        return value ? just(value) : nothing;
    },
    fromPredicate(pred, value) {
        switch (arguments.length) {
            case 1:
                return (value) => exports.Maybe.fromPredicate(pred, value);
            default:
                return pred(value) ? just(value) : nothing;
        }
    },
    mapMaybe(f, list) {
        switch (arguments.length) {
            case 1:
                return (list) => exports.Maybe.mapMaybe(f, list);
            default:
                return exports.Maybe.catMaybes(list.map(f));
        }
    },
    catMaybes(list) {
        let res = [];
        for (const e of list) {
            if (e.isJust()) {
                res.push(e.extract());
            }
        }
        return res;
    },
    encase(thunk) {
        try {
            return just(thunk());
        }
        catch (_a) {
            return nothing;
        }
    },
    isMaybe(x) {
        return x instanceof Just || x instanceof Nothing;
    },
    sequence(maybes) {
        let res = [];
        for (const m of maybes) {
            if (m.isJust()) {
                res.push(m.extract());
            }
            else {
                return nothing;
            }
        }
        return just(res);
    },
    'fantasy-land/of'(value) {
        return this.of(value);
    },
    'fantasy-land/empty'() {
        return this.empty();
    },
    'fantasy-land/zero'() {
        return this.zero();
    }
};
class Just {
    constructor(__value) {
        this.__value = __value;
    }
    isJust() {
        return true;
    }
    isNothing() {
        return false;
    }
    inspect() {
        return `Just(${this.__value})`;
    }
    [Symbol.for('nodejs.util.inspect.custom')](_depth, opts, inspect) {
        return `Just(${inspect(this.__value, opts)})`;
    }
    toString() {
        return this.inspect();
    }
    toJSON() {
        const value = this.__value;
        return value instanceof Date ? value.toJSON() : value;
    }
    equals(other) {
        return this.extract() === other.extract();
    }
    map(f) {
        return just(f(this.__value));
    }
    ap(maybeF) {
        return maybeF.isJust() ? this.map(maybeF.extract()) : nothing;
    }
    alt(_) {
        return this;
    }
    altLazy(_) {
        return this;
    }
    chain(f) {
        return f(this.__value);
    }
    chainNullable(f) {
        return exports.Maybe.fromNullable(f(this.__value));
    }
    join() {
        return this.__value;
    }
    reduce(reducer, initialValue) {
        return reducer(initialValue, this.__value);
    }
    extend(f) {
        return just(f(this));
    }
    unsafeCoerce() {
        return this.__value;
    }
    caseOf(patterns) {
        return '_' in patterns ? patterns._() : patterns.Just(this.__value);
    }
    orDefault(_) {
        return this.__value;
    }
    orDefaultLazy(_) {
        return this.__value;
    }
    toList() {
        return [this.__value];
    }
    mapOrDefault(f, _) {
        return f(this.__value);
    }
    extract() {
        return this.__value;
    }
    extractNullable() {
        return this.__value;
    }
    toEither(_) {
        return (0, Either_js_1.Right)(this.__value);
    }
    ifJust(effect) {
        return effect(this.__value), this;
    }
    ifNothing(_) {
        return this;
    }
    filter(pred) {
        return pred(this.__value) ? just(this.__value) : nothing;
    }
}
Just.prototype['fantasy-land/equals'] = Just.prototype.equals;
Just.prototype['fantasy-land/map'] = Just.prototype.map;
Just.prototype['fantasy-land/ap'] = Just.prototype.ap;
Just.prototype['fantasy-land/alt'] = Just.prototype.alt;
Just.prototype['fantasy-land/chain'] = Just.prototype.chain;
Just.prototype['fantasy-land/reduce'] = Just.prototype.reduce;
Just.prototype['fantasy-land/extend'] = Just.prototype.extend;
Just.prototype['fantasy-land/filter'] = Just.prototype.filter;
Just.prototype.constructor = exports.Maybe;
class Nothing {
    isJust() {
        return false;
    }
    isNothing() {
        return true;
    }
    inspect() {
        return 'Nothing';
    }
    [Symbol.for('nodejs.util.inspect.custom')]() {
        return 'Nothing';
    }
    toString() {
        return this.inspect();
    }
    toJSON() {
        return this.__value;
    }
    equals(other) {
        return this.extract() === other.extract();
    }
    map(_) {
        return nothing;
    }
    ap(_) {
        return nothing;
    }
    alt(other) {
        return other;
    }
    altLazy(other) {
        return other();
    }
    chain(_) {
        return nothing;
    }
    chainNullable(_) {
        return nothing;
    }
    join() {
        return nothing;
    }
    reduce(_, initialValue) {
        return initialValue;
    }
    extend(_) {
        return nothing;
    }
    unsafeCoerce() {
        throw new Error('Maybe#unsafeCoerce was ran on a Nothing');
    }
    caseOf(patterns) {
        return '_' in patterns ? patterns._() : patterns.Nothing();
    }
    orDefault(defaultValue) {
        return defaultValue;
    }
    orDefaultLazy(getDefaultValue) {
        return getDefaultValue();
    }
    toList() {
        return [];
    }
    mapOrDefault(_, defaultValue) {
        return defaultValue;
    }
    extract() {
        return undefined;
    }
    extractNullable() {
        return null;
    }
    toEither(left) {
        return (0, Either_js_1.Left)(left);
    }
    ifJust(_) {
        return this;
    }
    ifNothing(effect) {
        return effect(), this;
    }
    filter(_) {
        return nothing;
    }
}
Nothing.prototype['fantasy-land/equals'] = Nothing.prototype.equals;
Nothing.prototype['fantasy-land/map'] = Nothing.prototype.map;
Nothing.prototype['fantasy-land/ap'] = Nothing.prototype.ap;
Nothing.prototype['fantasy-land/alt'] = Nothing.prototype.alt;
Nothing.prototype['fantasy-land/chain'] = Nothing.prototype.chain;
Nothing.prototype['fantasy-land/reduce'] = Nothing.prototype.reduce;
Nothing.prototype['fantasy-land/extend'] = Nothing.prototype.extend;
Nothing.prototype['fantasy-land/filter'] = Nothing.prototype.filter;
Nothing.prototype.constructor = exports.Maybe;
/** Constructs a Just. Represents an optional value that exists */
const just = (value) => new Just(value);
exports.Just = just;
/** Represents a missing value, you can think of it as a smart 'null' */
const nothing = new Nothing();
exports.Nothing = nothing;
